Challenge2/src/agent.ts

```diff
@@ -0,0 +1,133 @@
+
+const getPoolValues = async (
+  provider: Provider,
+  uniPoolFunctionsAbi: string[],
+  poolAddress: string,
+  txEvent: TransactionEvent
+) => {
+  const poolContract = new ethers.Contract(
+    poolAddress,
+    uniPoolFunctionsAbi,
+    provider
+  );
+  const poolValues: any[] = await Promise.all([
+    poolContract.token0({ blockTag: txEvent.blockNumber }),
+    poolContract.token1({ blockTag: txEvent.blockNumber }),
+    poolContract.fee({ blockTag: txEvent.blockNumber }),
+  ]);
+  return poolValues;
+};
+
+export const getRealPoolAddress = async (
+  uniFactoryAddress: string,
+  uniInitCode: string,
+  interceptedPoolValues: any[]
+) => {
+  const interceptedPoolValuesBytes = ethers.utils.defaultAbiCoder.encode(
+    ["address", "address", "uint24"],
+    interceptedPoolValues
+  );
+  const interceptedSalt = ethers.utils.solidityKeccak256(
+    ["bytes"],
+    [interceptedPoolValuesBytes]
+  );
+  const realPoolAddress = ethers.utils.getCreate2Address(
+    uniFactoryAddress,
+    interceptedSalt,
+    uniInitCode
+  );
+  return realPoolAddress;
+};
+
+export function provideHandleTransaction(
+  uniFactoryAddress: string,
+  uniInitCode: string,
+  uniSwapEventAbi: string,
+  uniPoolFunctionsAbi: string[],
+  provider: Provider
+): HandleTransaction {
+  return async (txEvent: TransactionEvent) => {
+    const findings: Finding[] = [];
+    const filteredLogs = txEvent.filterLog(uniSwapEventAbi);
+
+    for (const filteredLog of filteredLogs) {
+      const interceptedPoolAddress = filteredLog.address;
+      const isPoolInCache = addressIsUniCache.get(interceptedPoolAddress);
+
+      if (isPoolInCache === false) return findings; // if was in cache and not a real pool, return
+
+      const interceptedPoolValues = await getPoolValues(
+        provider,
+        uniPoolFunctionsAbi,
+        interceptedPoolAddress,
+        txEvent
+      );
+
+      if (isPoolInCache === undefined) {
+        // if wan't in cache, check if it's a real pool
+        const realPoolAddress: string = await getRealPoolAddress(
+          uniFactoryAddress,
+          uniInitCode,
+          interceptedPoolValues
+        );
+        const isRealPool =
+          realPoolAddress.toLowerCase() ===
+          interceptedPoolAddress.toLowerCase();
+        addressIsUniCache.set(interceptedPoolAddress, isRealPool); // now it's in cache
+        if (!isRealPool) return findings; // if not a real pool, return
+      }
+
+      // remaining scenarios are: pool was in cache and is a real pool, or wan't in cache (but now it is) and is a real pool
+
+      const { sender, amount0, amount1, agentId } = filteredLog.args;
+
+      findings.push(
+        Finding.fromObject({
+          name: "Uniswap V3 Swap Detected",
+          description: `Address: ${sender} swapped ${amount0} ${interceptedPoolValues[0]} for ${amount1} ${interceptedPoolValues[1]}, using pool: ${interceptedPoolAddress}`,
```

Comment:
Perhaps you can use the absolute value for the `amount0` just to avoid the  "swapped -..."